lib/upperkut/strategies/priority_queue.rb

```diff
@@ -0,0 +1,186 @@
+module Upperkut
+  module Strategies
+    class PriorityQueue < Upperkut::Strategies::Base
+      include Upperkut::Util
+
+      # Logic as follows:
+      #
+      # We keep the last score used for each account key. One account_key is an account unique id.
+      #  to calculate the next_score we use max(current_account_score, current_global_score) + increment
+      #  we store the queue in a sorted set using the next_score as ordering key
+      #  if one account sends lots of messages, this account ends up with lots of
+      #  messages in the queue spaced by increment
+      #  if another account then sends a message, since it previous_account_score is lower than the
+      #  first account, it will be inserted before it in the queue
+      #
+      # In other words, the idea of this queue is to not allowing an account that sends a
+      #  lot of messages to dominate processing and give a chance for accounts that
+      #  sends few messages to have a fair share of processing time
+      ENQUEUE_ITEM = %(
+        local increment = 1
+        local current_checkpoint = tonumber(redis.call("GET", KEYS[1])) or 0
+        local account_score_key = KEYS[2]
+        local current_account_score = tonumber(redis.call("GET", account_score_key)) or 0
+        local queue_key = KEYS[3]
+        local next_score = nil
+
+        if current_account_score >= current_checkpoint then
+          next_score = current_account_score + increment
+        else
+          next_score = current_checkpoint + increment
+        end
+
+        redis.call("SETEX", account_score_key, 86400, next_score)
+        redis.call("ZADD", queue_key, next_score, ARGV[1])
+
+        return next_score
+      ).freeze
+
+      # Uses ZPOP* functions available only on redis 5.0.0+
```

Comment:
Deveriamos subir um erro antes executar o dequeue, quando o Redis não é suportado, faz sentido?

Reply:
Faz muito sentido mas pensamos em seguir sem isso no momento. Conseguimos ao menos forçar no gemspec a versão de um cliente do redis que suporta o Redis 5.0.0.